Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean

```diff
@@ -0,0 +1,201 @@
+  toFun := zmod n
+  map_zero_eq_one' := by simp
+  map_add_eq_mul' := by simp
+
+/-- Character on a product of `ZMod`s given by `x ↦ ∏ i, e ^ (2 * π * I * x i * y / n)`. -/
+private def mkZModAux {ι : Type} [DecidableEq ι] (n : ι → ℕ) [∀ i, NeZero (n i)]
+    (u : ∀ i, ZMod (n i)) : AddChar (⨁ i, ZMod (n i)) Circle :=
+  AddChar.directSum fun i ↦ zmod (n i) (u i)
+
+private lemma mkZModAux_injective {ι : Type} [DecidableEq ι] {n : ι → ℕ} [∀ i, NeZero (n i)] :
+    Injective (mkZModAux n) :=
+  AddChar.directSum_injective.comp fun f g h ↦ by simpa [funext_iff] using h
+
+/-- The circle-valued characters of a finite abelian group are the same as its complex-valued
+characters. -/
+def circleEquivComplex [Finite α] : AddChar α Circle ≃+ AddChar α ℂ where
+  toFun ψ := toMonoidHomEquiv.symm <| coeHom.comp ψ.toMonoidHom
+  invFun ψ :=
+    { toFun := fun a ↦ (⟨ψ a, mem_sphere_zero_iff_norm.2 <| ψ.norm_apply _⟩ : Circle)
+      map_zero_eq_one' := by simp [Circle]
+      map_add_eq_mul' := fun a b ↦ by ext : 1; simp [map_add_eq_mul] }
+  left_inv ψ := by ext : 1; simp
+  right_inv ψ := by ext : 1; simp
+  map_add' ψ χ := rfl
+
+@[simp] lemma card_eq [Fintype α] : card (AddChar α ℂ) = card α := by
+  obtain ⟨ι, _, n, hn, ⟨e⟩⟩ := AddCommGroup.equiv_directSum_zmod_of_finite' α
+  classical
+  have hn' i : NeZero (n i) := by have := hn i; exact ⟨by positivity⟩
+  let f : α → AddChar α ℂ := fun a ↦ coeHom.compAddChar ((mkZModAux n <| e a).compAddMonoidHom e)
+  have hf : Injective f := circleEquivComplex.injective.comp
+    ((compAddMonoidHom_injective_left _ e.surjective).comp <| mkZModAux_injective.comp <|
+      DFunLike.coe_injective.comp <| e.injective.comp Additive.ofMul.injective)
+  exact (card_addChar_le _ _).antisymm (Fintype.card_le_of_injective _ hf)
+
+/-- `ZMod n` is (noncanonically) isomorphic to its group of characters. -/
+def zmodAddEquiv : ZMod n ≃+ AddChar (ZMod n) ℂ := by
+  refine AddEquiv.ofBijective
+    (circleEquivComplex.toAddMonoidHom.comp <| AddChar.toAddMonoidHom zmodHom) ?_
+  rw [Fintype.bijective_iff_injective_and_card, card_eq]
+  exact ⟨circleEquivComplex.injective.comp zmod_injective, rfl⟩
+
+@[simp] lemma zmodAddEquiv_apply (x : ZMod n) :
+    zmodAddEquiv x = circleEquivComplex (zmod n x) := rfl
+
+section Finite
+variable (α) [Finite α]
+
+/-- Complex-valued characters of a finite abelian group `α` form a basis of `α → ℂ`. -/
+def complexBasis : Basis (AddChar α ℂ) ℂ (α → ℂ) :=
+  basisOfLinearIndependentOfCardEqFinrank (AddChar.linearIndependent _ _) <| by
+    cases nonempty_fintype α; rw [card_eq, Module.finrank_fintype_fun_eq_card]
+
+@[simp, norm_cast]
+lemma coe_complexBasis : ⇑(complexBasis α) = ((⇑) : AddChar α ℂ → α → ℂ) := by
+  rw [complexBasis, coe_basisOfLinearIndependentOfCardEqFinrank]
+
+variable {α}
+
+@[simp]
+lemma complexBasis_apply (ψ : AddChar α ℂ) : complexBasis α ψ = ψ := by rw [coe_complexBasis]
+
+lemma exists_apply_ne_zero : (∃ ψ : AddChar α ℂ, ψ a ≠ 1) ↔ a ≠ 0 := by
```

Comment:
It should be possible to reduce this to `CommGroup.exists_apply_ne_one_of_hasEnoughRootsOfUnity (Multiplicative α) ℂ`, but it looks like this requires building some API for `mulEquivMultiplicativeMonoidHom ` from above (and possibly variants).

So, on further thought, I think it is not worth the trouble, and it is OK to leave the code basically as is.